File: client/scripts/generate_docs_pdf.py
from __future__ import annotations

import re
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "PROJECT_DOCUMENTATION.md"
TARGET = ROOT / "PROJECT_DOCUMENTATION.pdf"

PAGE_WIDTH = 595
PAGE_HEIGHT = 842
LEFT_MARGIN = 48
RIGHT_MARGIN = 48
TOP_MARGIN = 52
BOTTOM_MARGIN = 52
LINE_GAP = 5
MAX_WIDTH = PAGE_WIDTH - LEFT_MARGIN - RIGHT_MARGIN


def escape_pdf_text(text: str) -> str:
    return text.replace("\\", "\\\\").replace("(", "\\(").replace(")", "\\)")


def estimate_text_width(text: str, font_size: int) -> float:
    return len(text) * font_size * 0.52


def wrap_text(text: str, font_size: int) -> list[str]:
    stripped = text.rstrip()
    if not stripped:
      return [""]

    words = stripped.split()
    lines: list[str] = []
    current = words[0]

    for word in words[1:]:
        trial = f"{current} {word}"
        if estimate_text_width(trial, font_size) <= MAX_WIDTH:
            current = trial
        else:
            lines.append(current)
            current = word

    lines.append(current)
    return lines


def tokenize_markdown(text: str) -> list[tuple[str, str]]:
    tokens: list[tuple[str, str]] = []
    in_code = False

    for raw_line in text.splitlines():
        line = raw_line.rstrip("\n")

        if line.startswith("```"):
            in_code = not in_code
            continue

        if in_code:
            tokens.append(("code", line))
            continue

        if not line.strip():
            tokens.append(("blank", ""))
            continue

        if line.startswith("### "):
            tokens.append(("h3", line[4:].strip()))
            continue

        if line.startswith("## "):
            tokens.append(("h2", line[3:].strip()))
            continue

        if line.startswith("# "):
            tokens.append(("h1", line[2:].strip()))
            continue

        if re.match(r"^\d+\.\s+", line):
            tokens.append(("list", line.strip()))
            continue

        if line.startswith("- "):
            tokens.append(("list", f"• {line[2:].strip()}"))
            continue

        tokens.append(("p", line))

    return tokens


def render_lines(tokens: list[tuple[str, str]]) -> list[tuple[str, int, str, int]]:
    rendered: list[tuple[str, int, str, int]] = []

    styles = {
        "h1": ("F2", 20, 18),
        "h2": ("F2", 14, 14),
        "h3": ("F2", 12, 12),
        "p": ("F1", 10, 9),
        "list": ("F1", 10, 9),
        "code": ("F3", 9, 9),
    }

    for kind, value in tokens:
        if kind == "blank":
            rendered.append(("blank", 0, "", 8))
            continue

        font_name, font_size, spacing = styles[kind]
        wrapped = wrap_text(value, font_size)
        for line in wrapped:
            rendered.append((font_name, font_size, line, spacing))

        if kind in {"h1", "h2", "h3"}:
            rendered.append(("blank", 0, "", 4))

    return rendered


def paginate(rendered_lines: list[tuple[str, int, str, int]]) -> list[list[tuple[str, int, str, int]]]:
    pages: list[list[tuple[str, int, str, int]]] = []
    current_page: list[tuple[str, int, str, int]] = []
    y = PAGE_HEIGHT - TOP_MARGIN

    for entry in rendered_lines:
        font_name, font_size, text, spacing = entry
        needed = spacing if font_name == "blank" else font_size + spacing
        if y - needed < BOTTOM_MARGIN:
            pages.append(current_page)
            current_page = []
            y = PAGE_HEIGHT - TOP_MARGIN

        current_page.append(entry)
        y -= needed

    if current_page:
        pages.append(current_page)

    return pages


def build_stream(page_lines: list[tuple[str, int, str, int]]) -> bytes:
    y = PAGE_HEIGHT - TOP_MARGIN
    chunks: list[str] = []

    for font_name, font_size, text, spacing in page_lines:
        if font_name == "blank":
            y -= spacing
            continue

        y -= font_size
        escaped = escape_pdf_text(text)
        chunks.append(f"BT /{font_name} {font_size} Tf 1 0 0 1 {LEFT_MARGIN} {y} Tm ({escaped}) Tj ET")
        y -= spacing

    return "\n".join(chunks).encode("latin-1", errors="replace")


def build_pdf(page_streams: list[bytes]) -> bytes:
    objects: list[bytes] = []

    def add_object(data: bytes) -> int:
        objects.append(data)
        return len(objects)

    font_regular = add_object(b"<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>")
    font_bold = add_object(b"<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica-Bold >>")
    font_code = add_object(b"<< /Type /Font /Subtype /Type1 /BaseFont /Courier >>")

    page_entries: list[tuple[int, int]] = []
    pages_placeholder = add_object(b"<<>>")

    for stream in page_streams:
        stream_obj = add_object(
            f"<< /Length {len(stream)} >>\nstream\n".encode("latin-1") + stream + b"\nendstream"
        )
        page_obj = add_object(
            (
                f"<< /Type /Page /Parent {pages_placeholder} 0 R "
                f"/MediaBox [0 0 {PAGE_WIDTH} {PAGE_HEIGHT}] "
                f"/Resources << /Font << /F1 {font_regular} 0 R /F2 {font_bold} 0 R /F3 {font_code} 0 R >> >> "
                f"/Contents {stream_obj} 0 R >>"
            ).encode("latin-1")
        )
        page_entries.append((stream_obj, page_obj))

    kids = " ".join(f"{page_obj} 0 R" for _, page_obj in page_entries)
    objects[pages_placeholder - 1] = (
        f"<< /Type /Pages /Count {len(page_entries)} /Kids [{kids}] >>".encode("latin-1")
    )

    catalog = add_object(f"<< /Type /Catalog /Pages {pages_placeholder} 0 R >>".encode("latin-1"))

    pdf = bytearray()
    pdf.extend(b"%PDF-1.4\n%\xe2\xe3\xcf\xd3\n")
    offsets = [0]

    for index, obj in enumerate(objects, start=1):
        offsets.append(len(pdf))
        pdf.extend(f"{index} 0 obj\n".encode("latin-1"))
        pdf.extend(obj)
        pdf.extend(b"\nendobj\n")

    xref_position = len(pdf)
    pdf.extend(f"xref\n0 {len(objects) + 1}\n".encode("latin-1"))
    pdf.extend(b"0000000000 65535 f \n")

    for offset in offsets[1:]:
        pdf.extend(f"{offset:010d} 00000 n \n".encode("latin-1"))

    pdf.extend(
        (
            f"trailer\n<< /Size {len(objects) + 1} /Root {catalog} 0 R >>\n"
            f"startxref\n{xref_position}\n%%EOF"
        ).encode("latin-1")
    )
    return bytes(pdf)


def main() -> None:
    markdown = SOURCE.read_text(encoding="utf-8")
    tokens = tokenize_markdown(markdown)
    rendered_lines = render_lines(tokens)
    pages = paginate(rendered_lines)
    streams = [build_stream(page) for page in pages]
    TARGET.write_bytes(build_pdf(streams))
    print(f"Created {TARGET}")


if __name__ == "__main__":
    main()

File: client/scripts/generate-docs-pdf.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const root = path.resolve(__dirname, '..');
const source = path.join(root, 'PROJECT_DOCUMENTATION.md');
const target = path.join(root, 'PROJECT_DOCUMENTATION.pdf');

const PAGE_WIDTH = 595;
const PAGE_HEIGHT = 842;
const LEFT_MARGIN = 48;
const RIGHT_MARGIN = 48;
const TOP_MARGIN = 52;
const BOTTOM_MARGIN = 52;
const MAX_WIDTH = PAGE_WIDTH - LEFT_MARGIN - RIGHT_MARGIN;

function escapePdfText(text) {
  return text.replace(/\\/g, '\\\\').replace(/\(/g, '\\(').replace(/\)/g, '\\)');
}

function estimateTextWidth(text, fontSize) {
  return text.length * fontSize * 0.52;
}

function wrapText(text, fontSize) {
  const stripped = text.trimEnd();
  if (!stripped) return [''];

  const words = stripped.split(/\s+/);
  const lines = [];
  let current = words[0];

  for (const word of words.slice(1)) {
    const trial = `${current} ${word}`;
    if (estimateTextWidth(trial, fontSize) <= MAX_WIDTH) {
      current = trial;
    } else {
      lines.push(current);
      current = word;
    }
  }

  lines.push(current);
  return lines;
}

function tokenizeMarkdown(text) {
  const tokens = [];
  let inCode = false;

  for (const rawLine of text.split(/\r?\n/)) {
    const line = rawLine;

    if (line.startsWith('```')) {
      inCode = !inCode;
      continue;
    }

    if (inCode) {
      tokens.push(['code', line]);
      continue;
    }

    if (!line.trim()) {
      tokens.push(['blank', '']);
      continue;
    }

    if (line.startsWith('### ')) {
      tokens.push(['h3', line.slice(4).trim()]);
      continue;
    }

    if (line.startsWith('## ')) {
      tokens.push(['h2', line.slice(3).trim()]);
      continue;
    }

    if (line.startsWith('# ')) {
      tokens.push(['h1', line.slice(2).trim()]);
      continue;
    }

    if (/^\d+\.\s+/.test(line)) {
      tokens.push(['list', line.trim()]);
      continue;
    }

    if (line.startsWith('- ')) {
      tokens.push(['list', `• ${line.slice(2).trim()}`]);
      continue;
    }

    tokens.push(['p', line]);
  }

  return tokens;
}

function renderLines(tokens) {
  const rendered = [];
  const styles = {
    h1: ['F2', 20, 18],
    h2: ['F2', 14, 14],
    h3: ['F2', 12, 12],
    p: ['F1', 10, 9],
    list: ['F1', 10, 9],
    code: ['F3', 9, 9],
  };

  for (const [kind, value] of tokens) {
    if (kind === 'blank') {
      rendered.push(['blank', 0, '', 8]);
      continue;
    }

    const [fontName, fontSize, spacing] = styles[kind];
    for (const line of wrapText(value, fontSize)) {
      rendered.push([fontName, fontSize, line, spacing]);
    }

    if (kind === 'h1' || kind === 'h2' || kind === 'h3') {
      rendered.push(['blank', 0, '', 4]);
    }
  }

  return rendered;
}

function paginate(renderedLines) {
  const pages = [];
  let currentPage = [];
  let y = PAGE_HEIGHT - TOP_MARGIN;

  for (const entry of renderedLines) {
    const [fontName, fontSize, , spacing] = entry;
    const needed = fontName === 'blank' ? spacing : fontSize + spacing;

    if (y - needed < BOTTOM_MARGIN) {
      pages.push(currentPage);
      currentPage = [];
      y = PAGE_HEIGHT - TOP_MARGIN;
    }

    currentPage.push(entry);
    y -= needed;
  }

  if (currentPage.length) pages.push(currentPage);
  return pages;
}

function buildStream(pageLines) {
  let y = PAGE_HEIGHT - TOP_MARGIN;
  const chunks = [];

  for (const [fontName, fontSize, text, spacing] of pageLines) {
    if (fontName === 'blank') {
      y -= spacing;
      continue;
    }

    y -= fontSize;
    const escaped = escapePdfText(text);
    chunks.push(`BT /${fontName} ${fontSize} Tf 1 0 0 1 ${LEFT_MARGIN} ${y} Tm (${escaped}) Tj ET`);
    y -= spacing;
  }

  return Buffer.from(chunks.join('\n'), 'latin1');
}

function buildPdf(pageStreams) {
  const objects = [];

  const addObject = (buffer) => {
    objects.push(buffer);
    return objects.length;
  };

  const fontRegular = addObject(Buffer.from('<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>'));
  const fontBold = addObject(Buffer.from('<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica-Bold >>'));
  const fontCode = addObject(Buffer.from('<< /Type /Font /Subtype /Type1 /BaseFont /Courier >>'));
  const pagesPlaceholder = addObject(Buffer.from('<<>>'));

  const pageEntries = [];

  for (const stream of pageStreams) {
    const streamObj = addObject(
      Buffer.concat([
        Buffer.from(`<< /Length ${stream.length} >>\nstream\n`, 'latin1'),
        stream,
        Buffer.from('\nendstream', 'latin1'),
      ]),
    );

    const pageObj = addObject(
      Buffer.from(
        `<< /Type /Page /Parent ${pagesPlaceholder} 0 R /MediaBox [0 0 ${PAGE_WIDTH} ${PAGE_HEIGHT}] /Resources << /Font << /F1 ${fontRegular} 0 R /F2 ${fontBold} 0 R /F3 ${fontCode} 0 R >> >> /Contents ${streamObj} 0 R >>`,
        'latin1',
      ),
    );

    pageEntries.push(pageObj);
  }

  objects[pagesPlaceholder - 1] = Buffer.from(
    `<< /Type /Pages /Count ${pageEntries.length} /Kids [${pageEntries.map((id) => `${id} 0 R`).join(' ')}] >>`,
    'latin1',
  );

  const catalog = addObject(Buffer.from(`<< /Type /Catalog /Pages ${pagesPlaceholder} 0 R >>`, 'latin1'));

  const pdf = [];
  pdf.push(Buffer.from('%PDF-1.4\n%\xE2\xE3\xCF\xD3\n', 'binary'));

  const offsets = [0];
  let currentLength = pdf[0].length;

  objects.forEach((obj, index) => {
    offsets.push(currentLength);
    const block = Buffer.concat([
      Buffer.from(`${index + 1} 0 obj\n`, 'latin1'),
      obj,
      Buffer.from('\nendobj\n', 'latin1'),
    ]);
    pdf.push(block);
    currentLength += block.length;
  });

  const xrefPosition = currentLength;
  let xref = `xref\n0 ${objects.length + 1}\n0000000000 65535 f \n`;
  for (const offset of offsets.slice(1)) {
    xref += `${String(offset).padStart(10, '0')} 00000 n \n`;
  }

  pdf.push(Buffer.from(xref, 'latin1'));
  pdf.push(
    Buffer.from(
      `trailer\n<< /Size ${objects.length + 1} /Root ${catalog} 0 R >>\nstartxref\n${xrefPosition}\n%%EOF`,
      'latin1',
    ),
  );

  return Buffer.concat(pdf);
}

const markdown = fs.readFileSync(source, 'utf8');
const tokens = tokenizeMarkdown(markdown);
const renderedLines = renderLines(tokens);
const pages = paginate(renderedLines);
const streams = pages.map(buildStream);

fs.writeFileSync(target, buildPdf(streams));
console.log(`Created ${target}`);
